aiidalab_widgets_base/viewers.py

```diff
@@ -335,9 +392,344 @@ def change_camera(change):
+            self.all_representation_output.children = []
+
+    def update_representations(self, change=None):
+        """Update the representations using the list of representations"""
+        number_of_representation_widgets = len(self.all_representations)
+        if self.displayed_structure:
+            if number_of_representation_widgets == 0:
+                self.n_all_representations = 0
+                self.add_representation(None)
+
+            representations = self.structure.arrays["representations"]
+            for rep in set(representations):
+                if (
+                    rep >= 0
+                ):  # negative values are used for atoms not represented (different from the case of hidden representations)
+                    self.all_representations[
+                        int(rep)
+                    ].selection.value = list_to_string_range(
+                        [int(i) for i in np.where(representations == rep)[0]], shift=1
+                    )
+            # empty selection field for unused representations
+            for rep in range(number_of_representation_widgets):
+                if rep not in {int(i) for i in representations}:
+                    self.all_representations[rep].selection.value = ""
+            self.apply_representations()
+
+    def representation_parameters(self, representation):
+        """Return the parameters dictionary of a representation."""
+        idsl = string_range_to_list(representation.selection.value, shift=-1)[0]
+        idsl_rep = [
+            i + rep * self.natoms
+            for rep in range(np.prod(self.supercell))
+            for i in idsl
+            if self.structure.arrays["representationsshow"][i]
+        ]
+        ids = self.list_to_nglview(idsl_rep)
+        radiusscale = 1
+        if representation.repr_type.value == "ball+stick":
+            radiusscale = 0.3
+        params = {
+            "type": "spacefill",
+            "params": {
+                "sele": ids,
+                "opacity": 1,
+                "color": representation.color.value,
+                "radiusScale": radiusscale * representation.radius.value,
+            },
+        }
+
+        return params
+
+    def compute_bonds(self, structure, radius=1.0, color="element", povray=False):
+        """Compute the bonds between atoms."""
+        radius = radius / 5
+
+        cutOff = neighborlist.natural_cutoffs(structure)
+        nl = neighborlist.NeighborList(cutOff, self_interaction=False, bothways=False)
+        nl.update(structure)
+        bonds = []
+        print("compute bonds")
+        if len(structure) > 1:
+            matrix = nl.get_connectivity_matrix()
+            for k in matrix.keys():
+                i = structure[k[0]]
+                j = structure[k[1]]
+
+                v1 = np.array([i.x, i.y, i.z])
+                v2 = np.array([j.x, j.y, j.z])
+                mic_vector = structure.get_distance(k[0], k[1], mic=True, vector=True)
```

Comment:
Here has two performance issues:
- you calculate the distance between atoms twice. 1) `nl.get_connectivity_matrix`; 2) `structure.get_distance`.
- you use a `for` loop, which is slow.

It takes around 1.3 s for 2000 atoms.

Please optimize this code. Feel free to have a discussion.

Reply:
not really sure how to improve this:
connectivity_matrix is the fastest way I know via ASE to get bonds in a structure after that
the loop runs only on bonded atoms (so is not N*N) and the get_distance is needed to have the MIC vector connecting two atoms needed to draw proper bonds also in the case atoms are at the edge of the cell. Do you know how to speed up this?

Reply:
wonderful @superstar54 I will try this way, thanks a lot!

Reply:
done :)